src/portfft/common/logging.hpp

```diff
@@ -261,6 +313,71 @@ struct global_data_struct {
   }
 };
 
+/*
+ * Outputs an object to std::cout. Most objects are piped directly to std::cout.
+ *
+ * @tparam T type of the object to output
+ * @param object object to output
+ */
+template <typename T, typename std::enable_if_t<!std::is_enum_v<T>>* = nullptr>
+__attribute__((always_inline)) inline void output(T object) {
+  std::cout << object;
+}
+
+/*
+ * Outputs an object to std::cout. Enums are first cast to underlying type.
+ *
+ * @tparam T type of the object to output
+ * @param object object to output
+ */
+template <typename T, typename std::enable_if_t<std::is_enum_v<T>>* = nullptr>
+__attribute__((always_inline)) inline void output(T object) {
+  output(static_cast<std::underlying_type_t<T>>(object));
+}
+
+/*
+ * Outputs an object to std::cout. A `std::vector` is output by elements.
+ *
+ * @tparam T type of the object to output
+ * @param object object to output
+ */
+template <typename T>
+__attribute__((always_inline)) inline void output(const std::vector<T>& object) {
+  std::cout << "(";
+  for (const T& element : object) {
+    output(element);
+    std::cout << ", ";
```

Comment:
It is simpler this way. Otherwise we would need to pass around a stringstream.

Reply:
I disagree because returning `std::string` from functions would enable
* Reuse in contexts that don't dump to std::cout
* Swapping out std::cout for something else
* Additionally, Finlay ([here](https://github.com/codeplaysoftware/portFFT/pull/135#discussion_r1469468680)) and I both initially expected a string to be returned. Outputting to `cout` seems less intuative. 

This is an opinion, so perhaps if anyone who agrees with this can thumbs-up. If that doesn't happen, go ahead and resolve this comment.

Reply:
My personal preference would be to output to a generic `std::ostream&` object as a first argument and return that same `ostream`. I think that's the most C++ like style.
I'm alright with the current solution too.

Reply:
In that case, LGTM.